File: packages/pulse/ts/src/wire.ts
/**
 * Wire codec — binary frame encode/decode. See spec/PROTOCOL.md §5.0 and
 * spec/FIXTURES.md. Pure functions; byte-for-byte identical to the Swift port.
 *
 * Layout (big-endian):
 *   header : u8 magic=0xB1 · u8 version=0x01 · u8 type
 *   str    : u8 len · len UTF-8 bytes
 *   blob   : u32 len · len bytes
 *   u64    : 8 bytes big-endian
 */

import type { Seq } from './types.js';

export const MAGIC = 0xb1;
export const VERSION = 0x01;

export const FrameType = {
  HELLO: 1,
  DATA: 2,
  ACK: 3,
  RESET: 4,
  HEARTBEAT: 5,
} as const;

export type Frame =
  | { t: 'hello'; epoch: string; recvEpoch: string; recvCursor: Seq }
  | { t: 'data'; seq: Seq; ack: Seq; payload: Uint8Array }
  | { t: 'ack'; ack: Seq }
  | { t: 'reset'; epoch: string; oldest: Seq }
  | { t: 'heartbeat'; ack: Seq };

const U64_MAX = (1n << 64n) - 1n;

// ── Encoder ──────────────────────────────────────────────────────────────────

class Writer {
  private parts: number[] = [];
  u8(n: number): void {
    this.parts.push(n & 0xff);
  }
  u32(n: number): void {
    this.parts.push((n >>> 24) & 0xff, (n >>> 16) & 0xff, (n >>> 8) & 0xff, n & 0xff);
  }
  u64(n: Seq): void {
    if (n < 0n || n > U64_MAX) throw new RangeError(`u64 out of range: ${n}`);
    for (let shift = 56n; shift >= 0n; shift -= 8n) {
      this.parts.push(Number((n >> shift) & 0xffn));
    }
  }
  str(s: string): void {
    const b = new TextEncoder().encode(s);
    if (b.length > 255) throw new RangeError(`str too long: ${b.length} bytes (max 255)`);
    this.u8(b.length);
    for (const x of b) this.parts.push(x);
  }
  blob(b: Uint8Array): void {
    if (b.length > 0xffff_ffff) throw new RangeError('blob too long');
    this.u32(b.length);
    for (const x of b) this.parts.push(x);
  }
  header(type: number): void {
    this.u8(MAGIC);
    this.u8(VERSION);
    this.u8(type);
  }
  done(): Uint8Array {
    return new Uint8Array(this.parts);
  }
}

export function encodeFrame(f: Frame): Uint8Array {
  const w = new Writer();
  switch (f.t) {
    case 'hello':
      w.header(FrameType.HELLO);
      w.str(f.epoch);
      w.str(f.recvEpoch);
      w.u64(f.recvCursor);
      break;
    case 'data':
      w.header(FrameType.DATA);
      w.u64(f.seq);
      w.u64(f.ack);
      w.blob(f.payload);
      break;
    case 'ack':
      w.header(FrameType.ACK);
      w.u64(f.ack);
      break;
    case 'reset':
      w.header(FrameType.RESET);
      w.str(f.epoch);
      w.u64(f.oldest);
      break;
    case 'heartbeat':
      w.header(FrameType.HEARTBEAT);
      w.u64(f.ack);
      break;
  }
  return w.done();
}

// ── Decoder ──────────────────────────────────────────────────────────────────

/** Bounds-checked cursor reader. Throws {@link Short} on underrun; the public
 *  decode function converts any throw into `null` (spec §5.0 robustness). */
class Short extends Error {}

class Reader {
  private off = 0;
  constructor(private readonly b: Uint8Array) {}
  private need(n: number): void {
    if (this.off + n > this.b.length) throw new Short();
  }
  u8(): number {
    this.need(1);
    return this.b[this.off++]!;
  }
  u32(): number {
    this.need(4);
    const v =
      (this.b[this.off]! * 0x1000000) +
      ((this.b[this.off + 1]! << 16) | (this.b[this.off + 2]! << 8) | this.b[this.off + 3]!);
    this.off += 4;
    return v >>> 0;
  }
  u64(): bigint {
    this.need(8);
    let v = 0n;
    for (let i = 0; i < 8; i++) v = (v << 8n) | BigInt(this.b[this.off + i]!);
    this.off += 8;
    return v;
  }
  str(): string {
    const len = this.u8();
    this.need(len);
    const slice = this.b.subarray(this.off, this.off + len);
    this.off += len;
    return new TextDecoder('utf-8', { fatal: false }).decode(slice);
  }
  blob(): Uint8Array {
    const len = this.u32();
    this.need(len);
    const slice = this.b.slice(this.off, this.off + len);
    this.off += len;
    return slice;
  }
  atEnd(): boolean {
    return this.off === this.b.length;
  }
}

/**
 * Decode wire bytes to a frame, or `null` if malformed / unknown / truncated.
 * MUST NOT throw on bad input.
 */
export function decodeFrame(bytes: Uint8Array): Frame | null {
  try {
    const r = new Reader(bytes);
    if (r.u8() !== MAGIC) return null;
    if (r.u8() !== VERSION) return null;
    const type = r.u8();
    switch (type) {
      case FrameType.HELLO: {
        const epoch = r.str();
        const recvEpoch = r.str();
        const recvCursor = r.u64();
        return { t: 'hello', epoch, recvEpoch, recvCursor };
      }
      case FrameType.DATA: {
        const seq = r.u64();
        const ack = r.u64();
        const payload = r.blob();
        return { t: 'data', seq, ack, payload };
      }
      case FrameType.ACK:
        return { t: 'ack', ack: r.u64() };
      case FrameType.RESET: {
        const epoch = r.str();
        const oldest = r.u64();
        return { t: 'reset', epoch, oldest };
      }
      case FrameType.HEARTBEAT:
        return { t: 'heartbeat', ack: r.u64() };
      default:
        return null; // unknown type
    }
  } catch {
    return null; // truncated / malformed
  }
}

File: packages/pulse/swift/Sources/Pulse/Wire.swift
import Foundation

/// Wire codec — binary frame encode/decode. See `spec/PROTOCOL.md` §5.0 and
/// `spec/FIXTURES.md`. Byte-for-byte identical to the TypeScript port.
///
/// Layout (big-endian):
///   header : u8 magic=0xB1 · u8 version=0x01 · u8 type
///   str    : u8 len · len UTF-8 bytes
///   blob   : u32 len · len bytes
///   u64    : 8 bytes big-endian

public let pulseMagic: UInt8 = 0xB1
public let pulseVersion: UInt8 = 0x01

public enum FrameType: UInt8 {
    case hello = 1
    case data = 2
    case ack = 3
    case reset = 4
    case heartbeat = 5
}

/// A decoded protocol frame.
public enum Frame: Equatable {
    case hello(epoch: String, recvEpoch: String, recvCursor: UInt64, durableSupported: Bool, maxRetentionMs: UInt64)
    case data(seq: UInt64, ack: UInt64, payload: [UInt8], durable: Bool)
    case ack(ack: UInt64)
    case reset(epoch: String, oldest: UInt64)
    case heartbeat(ack: UInt64)
}

public enum WireError: Error {
    case stringTooLong(Int)
}

// MARK: - Encoder

private struct Writer {
    var bytes: [UInt8] = []
    mutating func u8(_ n: UInt8) { bytes.append(n) }
    mutating func u32(_ n: UInt32) {
        bytes.append(UInt8((n >> 24) & 0xFF))
        bytes.append(UInt8((n >> 16) & 0xFF))
        bytes.append(UInt8((n >> 8) & 0xFF))
        bytes.append(UInt8(n & 0xFF))
    }
    mutating func u64(_ n: UInt64) {
        var shift: UInt64 = 56
        while true {
            bytes.append(UInt8((n >> shift) & 0xFF))
            if shift == 0 { break }
            shift -= 8
        }
    }
    mutating func str(_ s: String) throws {
        let b = Array(s.utf8)
        if b.count > 255 { throw WireError.stringTooLong(b.count) }
        u8(UInt8(b.count))
        bytes.append(contentsOf: b)
    }
    mutating func blob(_ b: [UInt8]) {
        u32(UInt32(b.count))
        bytes.append(contentsOf: b)
    }
    mutating func header(_ t: FrameType) {
        u8(pulseMagic)
        u8(pulseVersion)
        u8(t.rawValue)
    }
}

/// Encode a frame to its wire bytes.
public func encodeFrame(_ f: Frame) throws -> [UInt8] {
    var w = Writer()
    switch f {
    case let .hello(epoch, recvEpoch, recvCursor, durableSupported, maxRetentionMs):
        w.header(.hello)
        try w.str(epoch)
        try w.str(recvEpoch)
        w.u64(recvCursor)
        w.u8(durableSupported ? 1 : 0)
        w.u64(maxRetentionMs)
    case let .data(seq, ack, payload, durable):
        w.header(.data)
        w.u8(durable ? 1 : 0)
        w.u64(seq)
        w.u64(ack)
        w.blob(payload)
    case let .ack(ack):
        w.header(.ack)
        w.u64(ack)
    case let .reset(epoch, oldest):
        w.header(.reset)
        try w.str(epoch)
        w.u64(oldest)
    case let .heartbeat(ack):
        w.header(.heartbeat)
        w.u64(ack)
    }
    return w.bytes
}

// MARK: - Decoder

private struct Reader {
    let b: [UInt8]
    var off = 0
    init(_ bytes: [UInt8]) { b = bytes }

    mutating func need(_ n: Int) throws {
        if off + n > b.count { throw DecodeShort.short }
    }
    mutating func u8() throws -> UInt8 {
        try need(1)
        defer { off += 1 }
        return b[off]
    }
    mutating func u32() throws -> UInt32 {
        try need(4)
        defer { off += 4 }
        return (UInt32(b[off]) << 24) | (UInt32(b[off + 1]) << 16)
            | (UInt32(b[off + 2]) << 8) | UInt32(b[off + 3])
    }
    mutating func u64() throws -> UInt64 {
        try need(8)
        var v: UInt64 = 0
        for i in 0..<8 { v = (v << 8) | UInt64(b[off + i]) }
        off += 8
        return v
    }
    mutating func str() throws -> String {
        let len = Int(try u8())
        try need(len)
        let slice = Array(b[off..<off + len])
        off += len
        return String(decoding: slice, as: UTF8.self)
    }
    mutating func blob() throws -> [UInt8] {
        let len = Int(try u32())
        try need(len)
        let slice = Array(b[off..<off + len])
        off += len
        return slice
    }
}

private enum DecodeShort: Error { case short }

/// Decode wire bytes to a frame, or `nil` if malformed / unknown / truncated.
/// MUST NOT throw on bad input (spec §5.0 robustness).
public func decodeFrame(_ bytes: [UInt8]) -> Frame? {
    var r = Reader(bytes)
    do {
        if try r.u8() != pulseMagic { return nil }
        if try r.u8() != pulseVersion { return nil }
        let rawType = try r.u8()
        guard let type = FrameType(rawValue: rawType) else { return nil }
        switch type {
        case .hello:
            let epoch = try r.str()
            let recvEpoch = try r.str()
            let recvCursor = try r.u64()
            let durFlags = try r.u8()
            let maxRetentionMs = try r.u64()
            return .hello(
                epoch: epoch, recvEpoch: recvEpoch, recvCursor: recvCursor,
                durableSupported: (durFlags & 1) == 1, maxRetentionMs: maxRetentionMs)
        case .data:
            let msgFlags = try r.u8()
            let seq = try r.u64()
            let ack = try r.u64()
            let payload = try r.blob()
            return .data(seq: seq, ack: ack, payload: payload, durable: (msgFlags & 1) == 1)
        case .ack:
            return .ack(ack: try r.u64())
        case .reset:
            let epoch = try r.str()
            let oldest = try r.u64()
            return .reset(epoch: epoch, oldest: oldest)
        case .heartbeat:
            return .heartbeat(ack: try r.u64())
        }
    } catch {
        return nil
    }
}
